File: Davinci/logistical/logistical-android-api/src/main/java/io/bcteam/logistical/order/OrderRecordKeeper.java
package io.bcteam.logistical.order;

import io.bcteam.base.Request;
import io.bcteam.base.Response;
import io.bcteam.base.preconditionViolation.PreconditionViolation;
import io.bcteam.base.preconditionViolation.RequestNotValidException;
import io.bcteam.logistical.order.identifier.OrderIdentifier;
import io.bcteam.search.criteria.SearchCriteria;

import java.util.List;

public interface OrderRecordKeeper {
    CreateOrderResponse createOrder(
            CreateOrderRequest createOrderRequest) throws
            RequestNotValidException,
            OrderAlreadyExistsException;

    FindOrdersResponse findOrders(
            FindOrdersRequest findOrdersRequest) throws
            RequestNotValidException,
            UnsupportedSearchCriteriaException;

    RemoveOrderResponse removeOrder(
            RemoveOrderRequest removeOrderRequest) throws
            RequestNotValidException,
            UnsupportedOrderIdentifierException,
            OrderDoesNotExistException,
            OrderNotUniqueException;

    RetrieveOrderResponse retrieveOrder(
            RetrieveOrderRequest retrieveOrderRequest) throws
            RequestNotValidException,
            UnsupportedOrderIdentifierException,
            OrderNotUniqueException,
            OrderDoesNotExistException;

    UpdateOrderResponse updateOrder(
            UpdateOrderRequest updateOrderRequest) throws
            RequestNotValidException,
            UnsupportedOrderIdentifierException,
            OrderNotUniqueException,
            OrderDoesNotExistException;

    class CreateOrderResponse extends Response
    {
    }
    class CreateOrderRequest extends Request
    {
        private Order Order;

        public Order getOrder() {
            return Order;
        }

        public CreateOrderRequest setOrder(Order Order) {
            this.Order = Order;
            return this;
        }
    }

    class FindOrdersResponse extends Response
    {
        private List<Order> Orders;

        public List<Order> getOrders() {
            return Orders;
        }

        public FindOrdersResponse setOrders(List<Order> Orders) {
            this.Orders = Orders;
            return this;
        }
    }

    class FindOrdersRequest extends Request
    {
        private SearchCriteria searchCriteria;

        public SearchCriteria getSearchCriteria()
        {
            return searchCriteria;
        }

        public FindOrdersRequest setSearchCriteria(SearchCriteria searchCriteria)
        {
            this.searchCriteria = searchCriteria;
            return this;
        }
    }

    class UpdateOrderResponse extends Response
    {
    }

    class UpdateOrderRequest extends Request{

        private Order Order;

        private OrderIdentifier orderIdentifier;

        public Order getOrder()
        {
            return Order;
        }

        public UpdateOrderRequest setOrder(Order Order)
        {
            this.Order = Order;
            return this;
        }

        public OrderIdentifier getOrderIdentifier()
        {
            return orderIdentifier;
        }

        public UpdateOrderRequest setOrderIdentifier(OrderIdentifier orderIdentifier)
        {
            this.orderIdentifier = orderIdentifier;
            return this;
        }
    }

    class RemoveOrderResponse extends Response
    {

    }

    class RemoveOrderRequest extends Request
    {
        private OrderIdentifier OrderIdentifier;

        public OrderIdentifier getOrderIdentifier()
        {
            return OrderIdentifier;
        }

        public RemoveOrderRequest setOrderIdentifier(OrderIdentifier OrderIdentifier)
        {
            this.OrderIdentifier = OrderIdentifier;
            return this;
        }
    }

    class RetrieveOrderResponse extends Response
    {
        private Order Order;

        public Order getOrder() {
            return Order;
        }

        public RetrieveOrderResponse setOrder(Order Order) {
            this.Order = Order;
            return this;
        }
    }

    class RetrieveOrderRequest extends Request
    {
        private OrderIdentifier OrderIdentifier;

        public OrderIdentifier getOrderIdentifier() {
            return OrderIdentifier;
        }

        public RetrieveOrderRequest setOrderIdentifier(OrderIdentifier OrderIdentifier) {
            this.OrderIdentifier = OrderIdentifier;
            return this;
        }
    }

    class OrderAlreadyExistsException extends PreconditionViolation
    {
        public OrderAlreadyExistsException() {}

        public OrderAlreadyExistsException(String message)
        {
            super(message);
        }
    }

    class OrderDoesNotExistException extends PreconditionViolation
    {
        public OrderDoesNotExistException() {}

        public OrderDoesNotExistException(String message)
        {
            super(message);
        }
    }

    class UnsupportedOrderIdentifierException extends PreconditionViolation
    {

        public UnsupportedOrderIdentifierException(String message) {
            super(message);
        }

        public UnsupportedOrderIdentifierException() {}
    }

    class OrderNotUniqueException extends PreconditionViolation
    {
        public OrderNotUniqueException() {}

        public OrderNotUniqueException(String message)
        {
            super(message);
        }
    }

    class UnsupportedSearchCriteriaException extends PreconditionViolation
    {
        public UnsupportedSearchCriteriaException() {}

        public UnsupportedSearchCriteriaException(String message)
        {
            super(message);
        }
    }
}

File: SHSManagementSystem/SHSManagementSystem/BusinessLogicLayer/io/customerManagement/enquiries/order/OrderRecordKeeper.cs
using BusinessLayer.io.customerManagement.enquiries.order;
using BusinessLayer.io.fileHandler;
using BusinessLayer.io.globalExceptions;
using BusinessLayer.io.orderManagement.enquiries.order;
using BusinessLayer.io.orderManagement.enquiries.order.searchCriteria;
using BusinessLayer.io.search.criteria;
using DataAccessLayer.Persistance.UnitOfWork;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;

namespace BusinessLogicLayer.io.customerManagement.enquiries.order
{
    public class OrderRecordKeeper : IOrderRecordKeeper
    {
        private IUnitOfWork unitOfWork;
        private IFileHandler fileHandler;
        public OrderRecordKeeper(IUnitOfWork unitOfWork, IFileHandler fileHandler)
        {
            this.unitOfWork = unitOfWork;
            this.fileHandler = fileHandler;
        }
        public CreateOrderResponse CreateOrder(CreateOrderRequest createOrderRequest)
        {
            try
            {
                if (createOrderRequest.getOrder() == null)
                {
                    throw new RequestNotValid("CreateOrderRequest Not Valid.");
                }
                Order exceptionTest = RetrieveOrder(new RetrieveOrderRequest().setOrderTrackingNumber(
                    createOrderRequest.getOrder().TrackingNumber)).getOrder();

                if (exceptionTest != null)
                {
                    throw new OrderAlreadyExists("OrderAlreadyExists");
                }
                unitOfWork.CustomerEnquiries.Add(createOrderRequest.getOrder());
                unitOfWork.Complete();
            }
            catch (OrderAlreadyExists e)
            {
                return new CreateOrderResponse().setError(e.Message);
            }
            catch (RequestNotValid e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (Exception e)
            {
                fileHandler.AppendToTxt(new List<string>() { "Critical Error : " + e.Message });

            }
            return new CreateOrderResponse();
        }

        public FindOrderResponse FindOrder(FindOrderRequest findOrderRequest)
        {
            List<Order> orders = new List<Order>();
            try
            {
                if (findOrderRequest.getSearchCriteria() == null)
                {
                    throw new RequestNotValid("CreateOrderRequest Not Valid.");
                }

                List<Expression<Func<Order, object>>> orderIncluders = new List<Expression<Func<Order, object>>>();
                orderIncluders.Add(x => x.Customer);
                orderIncluders.Add(x => x.TechnicalTask);
                orderIncluders.Add(x => x.BillingInvoice);
                orderIncluders.Add(x => x.ProductsOrdered);

                if (findOrderRequest.getSearchCriteria() is AllSearch)
                {
                    orders = unitOfWork.Orders.GetAll(orderIncluders) as List<Order>;
                }
                else if (findOrderRequest.getSearchCriteria() is OrderSearchCriteria)
                {
                    string id = ((OrderSearchCriteria)findOrderRequest.getSearchCriteria()).getCustomerId();
                    orders = unitOfWork.Orders.Find(x => x.Customer.ID == id , orderIncluders) as List<Order>;
                }
                else
                {
                    throw new UnsupportedSearchCriteria("UnsupportedSearchCriteria");
                }

                if (orders.Count == 0)
                {
                    throw new OrderDoesNotExist("OrderDoesNotExist");
                }
            }
            catch (RequestNotValid e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (UnSupportedSearchIdentifier e)
            {
                return new FindOrderResponse().setError(e.Message);
            }
            catch (UnsupportedSearchCriteria e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (OrderDoesNotExist e)
            {
                return new FindOrderResponse().setError(e.Message);
            }
            catch (Exception e)
            {
                fileHandler.AppendToTxt(new List<string>() { "Critical error" + e.Message });
            }
            return new FindOrderResponse().setOrder(orders);
        }

        public RemoveOrderResponse RemoveOrder(RemoveOrderRequest removeOrderRequest)
        {
            try
            {
                if (removeOrderRequest.getOrder() == null)
                {
                    throw new RequestNotValid("RemoveOrderRequest Not Valid.");
                }
                Order exceptionTest = RetrieveOrder(new RetrieveOrderRequest().setOrderTrackingNumber(
                                         removeOrderRequest.getOrder().TrackingNumber)).getOrder();

                if (exceptionTest == null)
                {
                    throw new OrderDoesNotExist("OrderDoesNotExist");
                }
                unitOfWork.CustomerEnquiries.Remove(removeOrderRequest.getOrder());
                unitOfWork.Complete();
            }
            catch (RequestNotValid e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (OrderDoesNotExist e)
            {
                return new RemoveOrderResponse().setError(e.Message);
            }
            catch (Exception e)
            {
                fileHandler.AppendToTxt(new List<string>() { "Critical error" + e.Message });
            }
            return new RemoveOrderResponse();
        }

        public RetrieveOrderResponse RetrieveOrder(RetrieveOrderRequest retrieveOrderRequest)
        {
            Order order = null;
            try
            {
                if (retrieveOrderRequest.getOrderTrackingNumber() == null)
                {
                    throw new RequestNotValid("RetrieveOrderRequest Not Valid.");
                }
                List<Expression<Func<Order, object>>> orderIncluders = new List<Expression<Func<Order, object>>>();
                orderIncluders.Add(x => x.Customer);
                orderIncluders.Add(x => x.TechnicalTask);
                orderIncluders.Add(x => x.BillingInvoice);
                orderIncluders.Add(x => x.ProductsOrdered);

                order = unitOfWork.Orders.Get(retrieveOrderRequest.getOrderTrackingNumber(), orderIncluders);

                if (order == null)
                {
                    throw new OrderDoesNotExist("OrderDoesNotExist");
                }
            }
            catch (RequestNotValid e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (UnSupportedSearchIdentifier e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (OrderDoesNotExist e)
            {
                return new RetrieveOrderResponse().setError(e.Message);
            }
            catch (Exception e)
            {
                fileHandler.AppendToTxt(new List<string>() { "Critical error" + e.Message });
            }

            return new RetrieveOrderResponse().setOrder(order);
        }

        public UpdateOrderResponse UpdateOrder(UpdateOrderRequest updateOrderRequest)
        {
            Order order = null;
            try
            {
                if (updateOrderRequest.getOrder() == null)
                {
                    throw new RequestNotValid("UpdateOrderRequest Not Valid.");
                }

                order = RetrieveOrder(new RetrieveOrderRequest().setOrderTrackingNumber(
                                         updateOrderRequest.getOrder().TrackingNumber)).getOrder();

                if (order == null)
                {
                    throw new OrderDoesNotExist("OrderDoesNotExist");
                }

                order = updateOrderRequest.getOrder();
                unitOfWork.Complete();
            }
            catch (RequestNotValid e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (UnSupportedSearchIdentifier e)
            {
                fileHandler.AppendToTxt(new List<string>() { e.Message });
            }
            catch (OrderDoesNotExist e)
            {
                return new UpdateOrderResponse().setError(e.Message);
            }
            catch (Exception e)
            {
                fileHandler.AppendToTxt(new List<string>() { "Critical error" + e.Message });
            }
            return new UpdateOrderResponse().setOrder(order);
        }
    }
}
